src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/XmlKeyHelper.cs

```diff
@@ -0,0 +1,319 @@
+            Debug.Assert(name.Length > 0);
+            Debug.Assert(value.Length > 0);
+            Debug.Assert(builder != null);
+
+            builder.Append('<');
+            builder.Append(name);
+            builder.Append('>');
+
+            int offset = 0;
+            int length = value.Length;
+
+            // If we wanted to produce a ds:CryptoBinary instead of an xml:base64Binary,
+            // we'd skip all leading zeroes (increase offset, decrease length) before moving on
+
+            const int StackChars = 256;
+            const int ByteLimit = StackChars / 4 * 3;
+            Span<char> base64 = stackalloc char[StackChars];
+
+            while (length > 0)
+            {
+                int localLength = Math.Min(ByteLimit, length);
+
+                if (!Convert.TryToBase64Chars(value.Slice(offset, localLength), base64, out int written))
+                {
+                    Debug.Fail($"Convert.TryToBase64Chars failed with {localLength} bytes to {StackChars} chars");
+                    throw new CryptographicException();
+                }
+
+                builder.Append(base64.Slice(0, written));
+                length -= localLength;
+                offset += localLength;
+            }
+
+            builder.Append('<');
+            builder.Append('/');
+            builder.Append(name);
+            builder.Append('>');
+        }
+
+        internal struct ParseState
+        {
+            private IEnumerable _enumerable;
+            private IEnumerator _enumerator;
+            private int _index;
+
+            internal static ParseState ParseDocument(string xmlString)
+            {
+                object rootElement = Functions.ParseDocument(xmlString);
+
+                return new ParseState
+                {
+                    _enumerable = Functions.GetElements(rootElement),
+                    _enumerator = null,
+                    _index = -1,
+                };
+            }
+
+            internal bool HasElement(string localName)
+            {
+                string value = GetValue(localName);
+
+                bool ret = value != null;
+
+                if (ret)
+                {
+                    // Make it so that if GetValue is called on
+                    // this name it'll advance into it correctly.
+                    _index--;
+                }
+
+                return ret;
+            }
+
+            internal string GetValue(string localName)
+            {
+                if (_enumerable == null)
+                {
+                    return null;
+                }
+
+                if (_enumerator == null)
+                {
+                    _enumerator = _enumerable.GetEnumerator();
+                }
+
+                int origIdx = _index;
+                int idx = origIdx;
+
+                if (!_enumerator.MoveNext())
+                {
+                    idx = -1;
+                    _enumerator = _enumerable.GetEnumerator();
```

Comment:
So we loop through the whole enumerable twice?

Reply:
In happy degenerate cases: just once.  But every missing optional property requires checking all the values.  Basically I modelled it after how it was already working in NetFX; but I had thought (and am now thinking again) that a one-loop-into-a-switch might be better.